### ISSUE TITLE
Incorrectly getting out of gas exception.

### ISSUE THREAD

**Issue by AustP** [ConsenSys-archive/ganache-cli-archive#140]
## Expected Behavior

My contract has a maximum amount of ether that a participant can send to it. When someone sends more than the allowed amount, I expect the contract to keep the maximum and refund them the rest.
## Current Behavior

When sending more than the maximum amount of ether to my contract, I get `Error: VM Exception while executing transaction: out of gas(…)`.

The code block that handles the logic is as follows:

```
// if they go over the maximum, refund them the rest
if (participant.amount > maximum) {
    uint refund = participant.amount - maximum;
    participant.amount = maximum;
    amount -= refund;
    msg.sender.send(refund);
}
```

Now for the interesting bit. With a slightly tweaked code block, I can send more than the maximum amount of ether to my contract, and it works as expected... it takes the maximum amount and refunds them the rest. Here is the tweaked code block:

```
// if they go over the maximum, refund them the rest
if (participant.amount > maximum) {
    uint refund = participant.amount - maximum;
    amount -= refund;
    msg.sender.send(refund);


    // NOTICE: this line moved to the bottom
    participant.amount = maximum;
}
```

So for whatever reason, moving that line beneath the `send` method allows it work correctly.
## Your Environment
- Version used: 2.1.0
- Environment name and version: Truffle v2.0.4
- Operating System and version: xubuntu 16.04

**Comment by tcoulter** [comment]
Have you checked this against geth at all? Feels very much like a solidity compiler issue. If it fails on geth in the same way then we can be sure it's not an issue with the TestRPC.

Aside from that, giving us the full contract code and a descriptions of the transactions you made would be very helpful.

**Comment by cryptophonic** [comment]
FWIW - I have a similar issue with a contract that runs correctly on geth.  I'm trying to call a constant function that should not consume any gas, and works on geth, but am getting an out of gas exception on testrpc.

**Comment by cryptophonic** [comment]
Debugged: the constant function was throwing an error being interpreted as out of gas by testrpc while apparently geth fails silently...

**Comment by danfinlay** [comment]
I had a user at a hackathon getting this error incorrectly, after leaving testrpc running all day, and developing against it.  Restarting testrpc fixed the issue.

**Comment by tcoulter** [comment]
To be clear here, any `throw` in solidity code _is_ an OOG error - there is
no difference to the Ethereum virtual machine (this isn't a TestRPC issue;
it's the way Ethereum and Solidity were built). Considering that, when you
guys say "throwing an error", it's a bit confusing from my end. Ensure that
the code you're referring to has no `throw`'s, as that would in fact be a
correct OOG.

@danfinlay That doesn't necessarily give me enough to go on unfortunately.
Since a restart solves it, it would suggest to me it's a different, and
perhaps more subtle issue.

To everyone, full solidity code samples with transactions (in JS or JSON
RPC) would help immensely.


On Jan 28, 2017 10:47 PM, "Dan Finlay" <notifications@github.com> wrote:

I had a user at a hackathon getting this error incorrectly, after leaving
testrpc running all day, and developing against it. Restarting testrpc
fixed the issue.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/ethereumjs/testrpc/issues/140#issuecomment-275897486>,
or mute the thread

**Comment by illuzen** [comment]
I seem to be having a similar error. I have narrowed it down to a single line of code:

`myBool = true;`

Now myBool is public, and it is not used anywhere in the code except for here. Commenting it out makes everything run as expected. Leaving it in generates an out of gas exception, even when running with a ton of gas. Setting it to false doesn't change anything. Setting it to itself

`myBool = myBool`

does not generate the out of gas error. I've noticed this with several other variables. This only seems to happen when it's a contract calling another contract, not when I call it directly. 

I will test on geth as soon as I get my hands on some testether.
